File: platform/fastapi_server/routes/healthz_s3_cached_key.route.py
import json
import logging
import time

from aws_s3_client import config_from_env
from fastapi import FastAPI, Request
from healthz_diagnostics import ConfigSanitizer
from pydantic import BaseModel
from typing import Any, Optional

from cache_json_awss3_storage import (
    ClientAsync,
    create_storage,
    get_client_factory_from_app_config,
)

logger = logging.getLogger("fastapi_server.routes.healthz_s3_cached_key")


class OperationParams(BaseModel):
    key: Optional[str] = None
    data: Optional[dict[str, Any]] = None
    ttl: Optional[float] = None


class OperationRequest(BaseModel):
    config: Optional[dict[str, Any]] = None
    operation: str
    params: Optional[OperationParams] = None


def _get_yaml_s3(app: FastAPI) -> dict | None:
    """Get storage.s3 section from AppYamlConfig, or None if unavailable."""
    app_config = getattr(getattr(app, "state", None), "config", None)
    if app_config and hasattr(app_config, "get_nested"):
        return app_config.get_nested("storage", "s3")
    return None


def mount(app: FastAPI):
    """
    Mount S3 cached-key test routes.
    Provides a Postman-style API for testing cache_json_awss3_storage operations
    (save, load, delete, exists, list, clear) with user-supplied config.
    """

    @app.get("/healthz/s3-cached-key/defaults")
    async def healthz_s3_cached_key_defaults(request: Request):
        """Return server-resolved S3 config (credentials masked) for frontend pre-fill."""
        yaml_s3 = _get_yaml_s3(request.app)
        resolved = config_from_env(yaml_config=yaml_s3)
        return {
            "bucket_name": resolved.get("bucket_name") or "",
            "region_name": resolved.get("region") or "",
            "endpoint_url": resolved.get("endpoint_url") or "",
            "has_credentials": bool(
                resolved.get("aws_access_key_id") and resolved.get("aws_secret_access_key")
            ),
            "proxy_url": resolved.get("proxy_url") or "",
            "addressing_style": "path" if resolved.get("force_path_style") else "virtual",
            "key_prefix": resolved.get("key_prefix") or "jss3:",
        }

    @app.post("/healthz/s3-cached-key")
    async def healthz_s3_cached_key(request: Request):
        """
        Execute S3 cache operations with user-supplied or server-resolved config.

        Body:
            config: { bucket_name?, endpoint_url?, region_name?, aws_access_key_id?,
                      aws_secret_access_key?, proxy_url?, addressing_style?, verify?,
                      connection_timeout?, read_timeout?, key_prefix?, ttl? }
            operation: "save" | "load" | "delete" | "exists" | "list" | "clear"
            params: { key?, data?, ttl? }
        """
        body = await request.json()
        user_cfg = body.get("config") or {}
        operation = body.get("operation")
        params = body.get("params") or {}

        if not operation:
            return {"status": "error", "error": "operation is required"}

        # Build config: user overrides -> server defaults
        yaml_s3 = _get_yaml_s3(request.app)
        overrides: dict[str, Any] = {}
        if user_cfg.get("bucket_name"):
            overrides["bucket_name"] = user_cfg["bucket_name"]
        if user_cfg.get("region_name"):
            overrides["region_name"] = user_cfg["region_name"]
        if user_cfg.get("endpoint_url"):
            overrides["endpoint_url"] = user_cfg["endpoint_url"]
        if user_cfg.get("aws_access_key_id"):
            overrides["aws_access_key_id"] = user_cfg["aws_access_key_id"]
        if user_cfg.get("aws_secret_access_key"):
            overrides["aws_secret_access_key"] = user_cfg["aws_secret_access_key"]
        if user_cfg.get("proxy_url"):
            overrides["proxy_url"] = user_cfg["proxy_url"]
        if user_cfg.get("addressing_style"):
            overrides["addressing_style"] = user_cfg["addressing_style"]
        if user_cfg.get("connection_timeout") is not None:
            overrides["connection_timeout"] = user_cfg["connection_timeout"]
        if user_cfg.get("read_timeout") is not None:
            overrides["read_timeout"] = user_cfg["read_timeout"]
        if user_cfg.get("verify") is not None:
            overrides["verify"] = user_cfg["verify"]
        if user_cfg.get("ttl") is not None:
            overrides["ttl"] = user_cfg["ttl"]

        try:
            client_config = get_client_factory_from_app_config(yaml_s3, **overrides)
        except Exception as exc:
            return {
                "status": "error",
                "error": f"Config resolution failed: {exc}",
            }

        if not client_config.bucket_name:
            return {
                "status": "error",
                "error": "bucket_name is required (not in request or server config)",
            }

        key_prefix = user_cfg.get("key_prefix") or client_config.bucket_name and "jss3:" or "jss3:"
        default_ttl = client_config.ttl

        start = time.monotonic()
        try:
            async with ClientAsync(client_config) as s3_client:
                # Ensure bucket exists
                try:
                    await s3_client.head_bucket(Bucket=client_config.bucket_name)
                except Exception as bucket_err:
                    err_str = str(bucket_err).lower()
                    if "404" in err_str or "not found" in err_str or "nosuchbucket" in err_str or "403" in err_str:
                        logger.info("healthz/s3-cached-key: bucket not found, creating %s", client_config.bucket_name)
                        await s3_client.create_bucket(Bucket=client_config.bucket_name)
                        logger.info("healthz/s3-cached-key: bucket created %s", client_config.bucket_name)
                    else:
                        raise

                storage = create_storage(
                    s3_client,
                    client_config.bucket_name,
                    key_prefix=key_prefix,
                    ttl=default_ttl,
                )

                result = await _execute_operation(storage, operation, params, key_prefix, default_ttl, s3_client, client_config.bucket_name)
                latency_ms = round((time.monotonic() - start) * 1000)
                return {
                    "status": "ok",
                    "operation": operation,
                    "latency_ms": latency_ms,
                    "result": result,
                }

        except Exception as exc:
            latency_ms = round((time.monotonic() - start) * 1000)
            logger.error(
                "healthz/s3-cached-key: operation=%s failed (%dms) — %s",
                operation, latency_ms, str(exc),
            )
            return {
                "status": "error",
                "operation": operation,
                "latency_ms": latency_ms,
                "error": {
                    "name": type(exc).__name__,
                    "message": str(exc),
                    "code": getattr(exc, "code", None),
                    "cause": (
                        {
                            "name": type(exc.__cause__).__name__,
                            "message": str(exc.__cause__),
                            "code": getattr(exc.__cause__, "code", None),
                        }
                        if exc.__cause__
                        else None
                    ),
                },
            }


async def _execute_operation(
    storage,
    operation: str,
    params: dict,
    key_prefix: str,
    default_ttl: float | None,
    s3_client: Any,
    bucket_name: str,
) -> dict[str, Any]:
    """Dispatch to the appropriate storage operation."""

    if operation == "save":
        key = params.get("key")
        if not key:
            raise ValueError("params.key is required for save")
        data = params.get("data") or {}
        ttl = params.get("ttl") or default_ttl

        await storage.save(key, data, ttl=ttl)

        s3_key = f"{key_prefix}{key}"
        now = time.time()
        expires_at = now + ttl if ttl else None
        body_size = len(json.dumps({"key": key, "data": data, "created_at": now, "expires_at": expires_at}).encode("utf-8"))

        return {
            "key": key,
            "s3_key": s3_key,
            "ttl": ttl,
            "expires_at": expires_at,
            "size_bytes": body_size,
        }

    elif operation == "load":
        key = params.get("key")
        if not key:
            raise ValueError("params.key is required for load")

        s3_key = f"{key_prefix}{key}"

        # Use raw S3 GET to get full entry (including metadata)
        try:
            response = await s3_client.get_object(Bucket=bucket_name, Key=s3_key)
            body = response["Body"]
            content = await body.read() if hasattr(body, "read") else body
            entry = json.loads(content)

            now = time.time()
            is_expired = entry.get("expires_at") is not None and now > entry["expires_at"]

            return {
                "key": key,
                "s3_key": s3_key,
                "found": True,
                "expired": is_expired,
                "entry": entry,
            }
        except Exception as e:
            err_str = str(e).lower()
            if "nosuchkey" in err_str or "404" in err_str or "not found" in err_str:
                return {"key": key, "s3_key": s3_key, "found": False, "data": None}
            raise

    elif operation == "delete":
        key = params.get("key")
        if not key:
            raise ValueError("params.key is required for delete")

        await storage.delete(key)
        return {"key": key, "s3_key": f"{key_prefix}{key}", "deleted": True}

    elif operation == "exists":
        key = params.get("key")
        if not key:
            raise ValueError("params.key is required for exists")

        exists = await storage.exists(key)
        return {"key": key, "s3_key": f"{key_prefix}{key}", "exists": exists}

    elif operation == "list":
        keys = await storage.list_keys()
        key_details = []
        for k in keys:
            key_details.append({
                "key": k,
                "s3_key": f"{key_prefix}{k}",
            })
        return {
            "key_prefix": key_prefix,
            "count": len(keys),
            "keys": key_details,
        }

    elif operation == "clear":
        deleted_count = await storage.clear()
        return {
            "key_prefix": key_prefix,
            "deleted_count": deleted_count,
        }

    else:
        raise ValueError(
            f"Unknown operation: {operation}. Use: save, load, delete, exists, list, clear"
        )

File: platform/fastify_server/routes/healthz_s3_cached_key.route.mjs
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
  DeleteObjectCommand,
  HeadObjectCommand,
  ListObjectsV2Command,
  DeleteObjectsCommand,
  HeadBucketCommand,
  CreateBucketCommand,
} from "@aws-sdk/client-s3";
import { NodeHttpHandler } from "@smithy/node-http-handler";
import { configFromEnv } from "aws-s3-client";

/**
 * Build an S3Client from user-supplied config (Postman-style).
 * @param {object} cfg
 * @returns {S3Client}
 */
async function buildClientFromConfig(cfg) {
  const clientConfig = {};

  if (cfg.region_name) {
    clientConfig.region = cfg.region_name;
  }
  if (cfg.endpoint_url) {
    clientConfig.endpoint = cfg.endpoint_url;
    clientConfig.forcePathStyle = cfg.addressing_style === "path";
  }
  if (cfg.aws_access_key_id && cfg.aws_secret_access_key) {
    clientConfig.credentials = {
      accessKeyId: cfg.aws_access_key_id,
      secretAccessKey: cfg.aws_secret_access_key,
    };
  }
  if (cfg.verify === false) {
    clientConfig.tls = false;
  }
  if (cfg.proxy_url) {
    const { HttpsProxyAgent } = await import("https-proxy-agent");
    const agent = new HttpsProxyAgent(cfg.proxy_url);
    clientConfig.requestHandler = new NodeHttpHandler({
      httpAgent: agent,
      httpsAgent: agent,
      connectionTimeout: (cfg.connection_timeout || 20) * 1000,
      requestTimeout: (cfg.read_timeout || 60) * 1000,
    });
  } else {
    clientConfig.requestHandler = new NodeHttpHandler({
      connectionTimeout: (cfg.connection_timeout || 20) * 1000,
      requestTimeout: (cfg.read_timeout || 60) * 1000,
    });
  }

  return new S3Client(clientConfig);
}

/**
 * Mount S3 cached-key test routes.
 * Provides a Postman-style API for testing cache_json_awss3_storage operations
 * (save, load, delete, exists, list, clear) with user-supplied config.
 * @param {import('fastify').FastifyInstance} server
 */
export async function mount(server) {
  /**
   * Resolve server-side S3 config from YAML + env (same source as /healthz/s3).
   */
  function getServerConfig() {
    const yamlS3 = server.config?.getNested?.(['storage', 's3']) ?? null;
    return configFromEnv(undefined, yamlS3);
  }

  /**
   * GET /healthz/s3-cached-key/defaults
   *
   * Returns the server-resolved S3 config (credentials masked) so the
   * frontend can show defaults and pre-fill fields.
   */
  server.get("/healthz/s3-cached-key/defaults", async () => {
    const sc = getServerConfig();
    return {
      bucket_name: sc.bucketName || "",
      region_name: sc.region || "",
      endpoint_url: sc.endpointUrl || "",
      has_credentials: !!(sc.awsAccessKeyId && sc.awsSecretAccessKey),
      proxy_url: sc.proxyUrl || "",
      addressing_style: sc.forcePathStyle ? "path" : "virtual",
      key_prefix: sc.keyPrefix || "jss3:",
    };
  });

  /**
   * POST /healthz/s3-cached-key
   *
   * Body:
   *   config: { bucket_name?, endpoint_url?, region_name?, aws_access_key_id?,
   *             aws_secret_access_key?, proxy_url?, addressing_style?, verify?,
   *             connection_timeout?, read_timeout?, key_prefix?, ttl? }
   *   operation: "save" | "load" | "delete" | "exists" | "list" | "clear"
   *   params: { key?, data?, ttl? }
   *
   * Optional config fields fall back to server-resolved values
   * (YAML app config → environment variables).
   */
  server.post("/healthz/s3-cached-key", async (request, reply) => {
    const { config: userCfg = {}, operation, params = {} } = request.body || {};

    if (!operation) {
      reply.code(400);
      return { status: "error", error: "operation is required" };
    }

    // Merge: user-supplied values → server defaults
    const sc = getServerConfig();
    const cfg = {
      bucket_name: userCfg.bucket_name || sc.bucketName,
      endpoint_url: userCfg.endpoint_url || sc.endpointUrl,
      region_name: userCfg.region_name || sc.region,
      aws_access_key_id: userCfg.aws_access_key_id || sc.awsAccessKeyId,
      aws_secret_access_key: userCfg.aws_secret_access_key || sc.awsSecretAccessKey,
      proxy_url: userCfg.proxy_url || sc.proxyUrl,
      addressing_style: userCfg.addressing_style ?? (sc.forcePathStyle ? "path" : "virtual"),
      connection_timeout: userCfg.connection_timeout ?? 20,
      read_timeout: userCfg.read_timeout ?? 60,
      verify: userCfg.verify ?? true,
      key_prefix: userCfg.key_prefix ?? sc.keyPrefix ?? "jss3:",
      ttl: userCfg.ttl,
    };

    if (!cfg.bucket_name) {
      reply.code(400);
      return { status: "error", error: "bucket_name is required (not in request or server config)" };
    }

    const keyPrefix = cfg.key_prefix || "jss3:";
    const defaultTtl = cfg.ttl ?? null;
    const bucket = cfg.bucket_name;

    let client;
    const start = performance.now();
    try {
      client = await buildClientFromConfig(cfg);

      // Ensure bucket exists — create if missing
      try {
        await client.send(new HeadBucketCommand({ Bucket: bucket }));
      } catch (err) {
        const errName = (err.name || "").toLowerCase();
        const errCode = (err.$metadata?.httpStatusCode || 0);
        if (errName === "notfound" || errName === "nosuchbucket" || errCode === 404 || errCode === 403) {
          server.log.info({ bucket }, "healthz/s3-cached-key: bucket not found, creating");
          await client.send(new CreateBucketCommand({ Bucket: bucket }));
          server.log.info({ bucket }, "healthz/s3-cached-key: bucket created");
        } else {
          throw err;
        }
      }

      switch (operation) {
        // ── SAVE ──
        case "save": {
          const key = params.key;
          if (!key) {
            reply.code(400);
            return { status: "error", error: "params.key is required for save" };
          }
          const data = params.data ?? {};
          const ttl = params.ttl ?? defaultTtl;
          const now = Date.now() / 1000;
          const expiresAt = ttl ? now + ttl : null;

          const entry = {
            key,
            data,
            created_at: now,
            expires_at: expiresAt,
          };

          const s3Key = `${keyPrefix}${key}`;
          const body = JSON.stringify(entry);

          await client.send(
            new PutObjectCommand({
              Bucket: bucket,
              Key: s3Key,
              Body: body,
              ContentType: "application/json",
            })
          );

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "save",
            latency_ms,
            result: {
              key,
              s3_key: s3Key,
              ttl: ttl ?? null,
              expires_at: expiresAt,
              size_bytes: Buffer.byteLength(body, "utf-8"),
            },
          };
        }

        // ── LOAD ──
        case "load": {
          const key = params.key;
          if (!key) {
            reply.code(400);
            return { status: "error", error: "params.key is required for load" };
          }
          const s3Key = `${keyPrefix}${key}`;

          let response;
          try {
            response = await client.send(
              new GetObjectCommand({ Bucket: bucket, Key: s3Key })
            );
          } catch (err) {
            const errStr = String(err).toLowerCase();
            if (
              errStr.includes("nosuchkey") ||
              errStr.includes("404") ||
              errStr.includes("not found")
            ) {
              const latency_ms = Math.round(performance.now() - start);
              return {
                status: "ok",
                operation: "load",
                latency_ms,
                result: { key, s3_key: s3Key, found: false, data: null },
              };
            }
            throw err;
          }

          const content = await response.Body.transformToString();
          let entry;
          try {
            entry = JSON.parse(content);
          } catch {
            entry = { raw: content };
          }

          // Check TTL
          const now = Date.now() / 1000;
          const isExpired =
            entry.expires_at != null && now > entry.expires_at;

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "load",
            latency_ms,
            result: {
              key,
              s3_key: s3Key,
              found: true,
              expired: isExpired,
              entry,
            },
          };
        }

        // ── DELETE ──
        case "delete": {
          const key = params.key;
          if (!key) {
            reply.code(400);
            return { status: "error", error: "params.key is required for delete" };
          }
          const s3Key = `${keyPrefix}${key}`;

          await client.send(
            new DeleteObjectCommand({ Bucket: bucket, Key: s3Key })
          );

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "delete",
            latency_ms,
            result: { key, s3_key: s3Key, deleted: true },
          };
        }

        // ── EXISTS ──
        case "exists": {
          const key = params.key;
          if (!key) {
            reply.code(400);
            return { status: "error", error: "params.key is required for exists" };
          }
          const s3Key = `${keyPrefix}${key}`;

          let exists = false;
          try {
            await client.send(
              new HeadObjectCommand({ Bucket: bucket, Key: s3Key })
            );
            exists = true;
          } catch (err) {
            const errStr = String(err).toLowerCase();
            if (
              errStr.includes("404") ||
              errStr.includes("not found") ||
              errStr.includes("nosuchkey")
            ) {
              exists = false;
            } else {
              throw err;
            }
          }

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "exists",
            latency_ms,
            result: { key, s3_key: s3Key, exists },
          };
        }

        // ── LIST ──
        case "list": {
          const allKeys = [];
          let continuationToken;

          while (true) {
            const listParams = {
              Bucket: bucket,
              Prefix: keyPrefix,
              MaxKeys: 100,
            };
            if (continuationToken) {
              listParams.ContinuationToken = continuationToken;
            }

            const response = await client.send(
              new ListObjectsV2Command(listParams)
            );

            for (const obj of response.Contents || []) {
              const s3Key = obj.Key || "";
              const storageKey = s3Key.startsWith(keyPrefix)
                ? s3Key.slice(keyPrefix.length)
                : s3Key;
              allKeys.push({
                key: storageKey,
                s3_key: s3Key,
                size: obj.Size,
                last_modified: obj.LastModified?.toISOString() || null,
              });
            }

            if (response.IsTruncated) {
              continuationToken = response.NextContinuationToken;
            } else {
              break;
            }
          }

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "list",
            latency_ms,
            result: { key_prefix: keyPrefix, count: allKeys.length, keys: allKeys },
          };
        }

        // ── CLEAR ──
        case "clear": {
          // List all keys with prefix, then batch-delete
          const keysToDelete = [];
          let continuationToken;

          while (true) {
            const listParams = {
              Bucket: bucket,
              Prefix: keyPrefix,
              MaxKeys: 1000,
            };
            if (continuationToken) {
              listParams.ContinuationToken = continuationToken;
            }

            const response = await client.send(
              new ListObjectsV2Command(listParams)
            );

            for (const obj of response.Contents || []) {
              keysToDelete.push({ Key: obj.Key });
            }

            if (response.IsTruncated) {
              continuationToken = response.NextContinuationToken;
            } else {
              break;
            }
          }

          if (keysToDelete.length > 0) {
            // Delete in batches of 1000
            for (let i = 0; i < keysToDelete.length; i += 1000) {
              const batch = keysToDelete.slice(i, i + 1000);
              await client.send(
                new DeleteObjectsCommand({
                  Bucket: bucket,
                  Delete: { Objects: batch },
                })
              );
            }
          }

          const latency_ms = Math.round(performance.now() - start);
          return {
            status: "ok",
            operation: "clear",
            latency_ms,
            result: {
              key_prefix: keyPrefix,
              deleted_count: keysToDelete.length,
            },
          };
        }

        default:
          reply.code(400);
          return {
            status: "error",
            error: `Unknown operation: ${operation}. Use: save, load, delete, exists, list, clear`,
          };
      }
    } catch (err) {
      const latency_ms = Math.round(performance.now() - start);
      server.log.error(
        { operation, latency_ms, error: err.message },
        "healthz/s3-cached-key: operation failed"
      );
      reply.code(500);
      return {
        status: "error",
        operation,
        latency_ms,
        error: {
          name: err.name || "Error",
          message: err.message,
          code: err.code || null,
        },
      };
    } finally {
      if (client) {
        try {
          client.destroy();
        } catch (_) {
          // ignore
        }
      }
    }
  });
}
